src/apps/vlan/vlan.lua

```diff
@@ -0,0 +1,174 @@
+module(..., package.seeall)
+
+local packet = require("core.packet")
+local bit = require("bit")
+local ffi = require("ffi")
+
+local C = ffi.C
+local receive, transmit = link.receive, link.transmit
+local cast = ffi.cast
+
+Tagger = {}
+Untagger = {}
+
+-- 802.1q
+local dotq_tpid = 0x8100
+local o_ethernet_ethertype = 12
+local uint32_ptr_t = ffi.typeof('uint32_t*')
+
+
+-- build a VLAN tag consisting of 2 bytes of TPID set to 0x8100 followed by the
+-- TCI field which in turns consists of PCP, DEI and VID (VLAN id). Both PCP
+-- and DEI is always 0
+local function build_tag(vid)
+   return ffi.C.htonl(bit.bor(bit.lshift(dotq_tpid, 16), vid))
+end
+
+-- pop a VLAN tag (4 byte of TPID and TCI) from a packet
+function pop_tag(pkt)
+   local payload = pkt.data + o_ethernet_ethertype
+   local length = pkt.length
+   pkt.length = length - 4
+   C.memmove(payload, payload + 4, length - o_ethernet_ethertype - 4)
+end
+
+-- push a VLAN tag onto a packet
+function push_tag(pkt, tag)
+   local payload = pkt.data + o_ethernet_ethertype
+   local length = pkt.length
+   pkt.length = length + 4
+   C.memmove(payload + 4, payload, length - o_ethernet_ethertype)
+   cast(uint32_ptr_t, payload)[0] = tag
+end
+
+-- extract TCI (2 bytes) from packet, no check is performed to verify that the
+-- packet is carrying a VLAN tag, if it's an untagged frame these bytes will be
+-- Ethernet payload
+function extract_tci(pkt)
+   return ffi.C.ntohs(ffi.cast("uint16_t*", packet.data(pkt) + o_ethernet_ethertype + 2)[0])
```

Comment:
`cast`is cached.

There's the functions `rd16, `rd32`, `w16`, `w32` in the lwAFTR codebase (`apps/lwaftr/lwutil.lua`) which do the same (casting of a pointer and derefence) but makes code easier to read. I understand that importing a lwAFTR module maybe understood as if this module was dependent on the lwAFTR. I think ideally these helper functions should be in `core/lib.lua` but that's something to do in another PR IMHO. In conclusion, I would leave as it is for now but I would love to have it changed in the future.